File: src/app/utils/gcode.ts
export class GCodeLine {
    private _cmd = '';
    private _x: number | null = null;
    private _y: number | null = null;
    private _z: number | null = null;
    private _f: number | null = null;
    private _s: number | null = null;
    private readonly raw: string;
  
    constructor(raw: string) {
      this.raw = raw;
      const parts = raw.split(' ');
      this._cmd = parts[0];
      parts.slice(1).forEach((_part) => {
        // Ignore anything after a comment
        const part = _part.split(';')[0];
        switch (part[0].toUpperCase()) {
          case 'X':
            this._x = Number(part.substring(1));
            break;
          case 'Y':
            this._y = Number(part.substring(1));
            break;
          case 'Z':
            this._z = Number(part.substring(1));
            break;
          case 'F':
            this._f = Number(part.substring(1));
            break;
            case 'S':
                this._s = Number(part.substring(1));
                break;
        }
      });
    }
  
    get cmd() {
      return this._cmd;
    }
  
    get x() {
      return this._x;
    }
  
    get y() {
      return this._y;
    }
  
    get z() {
      return this._z;
    }
  
    set z(val: number) {
      this._z = val;
    }
  
    get f() {
      return this._f;
    }
  
    repr() {
      const xStr = this._x || this._x === 0 ? ` X${this._x.toFixed(8)}` : '';
      const yStr = this._y || this._y === 0 ? ` Y${this._y.toFixed(8)}` : '';
      const zStr = this._z || this._z === 0 ? ` Z${this._z.toFixed(8)}` : '';
      const fStr = this._f || this._f === 0 ? ` F${this._f.toFixed(8)}` : '';
      const sStr = this._s || this._s === 0 ? ` F${this._s.toFixed(8)}` : '';
      return `${this._cmd}${xStr}${yStr}${zStr}${fStr}${sStr}`;
    }
  }
  
  export class GCodeObject {
    rawGcodeOps: string[] = [];
    gcodeLines: GCodeLine[] = [];
  
    constructor(gcode: string) {
      this.rawGcodeOps = gcode
        .split('\n')
        .map((x) => x.trim())
        .filter((x) => {
          return x.length > 0 && !x.startsWith('(') && !x.startsWith(';');
        });
      this.gcodeLines = this.rawGcodeOps.map((x) => new GCodeLine(x));
    }
  }

File: py/gcode.py
import io
from typing import TextIO, Union

class GCodeLine:
    cmd = ''
    x = None
    y = None
    z = None
    f = None
    s = None
    __raw = ''

    def __init__(self, raw: str):
        self.__raw = raw
        parts = raw.split(' ')
        self.cmd = parts[0]
        for part in parts[1:]:
            part = part.split(';')[0]
            if part[0].upper() == 'X':
                self.x = float(part[1:])
            elif part[0].upper() == 'Y':
                self.y = float(part[1:])
            elif part[0].upper() == 'Z':
                self.z = float(part[1:])
            elif part[0].upper() == 'F':
                self.f = float(part[1:])
            elif part[0].upper() == 'S':
                self.s = float(part[1:])

    def __repr__(self):
        x_str = '' if self.x is None else f' X{self.x:.8f}'
        y_str = '' if self.y is None else f' Y{self.y:.8f}'
        z_str = '' if self.z is None else f' Z{self.z:.8f}'
        f_str = '' if self.f is None else f' F{self.f:.8f}'
        s_str = '' if self.s is None else f' S{self.s:.8f}'
        return f'{self.cmd}{x_str}{y_str}{z_str}{f_str}{s_str}\n'

class GCodeObject:

    raw_gcode_ops: list[str] = []
    gcode_lines: list[GCodeLine] = []
    
    def __init__(self, file: Union[str, TextIO]):
        lines = []
        if type(file) is str:
            lines = file.splitlines()
        else:
            with open(file) as f:
                lines = f.readlines()
        # Filter empty lines and G Code comments
        self.raw_gcode_ops = list(filter(lambda y: len(y) > 0 and not y.startswith('(') and not y.startswith(';'), [x.strip() for x in lines]))
        self.gcode_lines = [GCodeLine(x) for x in self.raw_gcode_ops]

    def create_file(self, name: str):
        with open(name, 'w') as f:
            f.writelines(map(str, self.gcode_lines))

    def write_to_stream(self, stream: io.TextIOBase):
        stream.writelines(map(str, self.gcode_lines))

if __name__ == '__main__':
    GCodeObject('example.gcode')